### PR TITLE
[Java] Choose The Correct RoundingMode For Checking Decimal OutOfBounds

### LINKED ISSUE #14732: [BUG] Error In Calculating Bounds for DecimalUtils OOB Check
**Describe the bug**
The `DecimalUtils.lessThan` and `DecimalUtils.greaterThan` methods don't consider whether the RHS value is negative or positive before deciding if the value should be rounded up or down. Leading to errors while converting scales

**Steps/Code to reproduce bug**

The value `-1E+3` is out of bounds for precision 1 and scale 1 but will not be reported as such and the following test will fail. 

For precision = 1 and scale = 1 the upper/lower bounds are 90/-90 but during conversion, the upper/lower bounds become 0/-1000 causing the test value to fall wrongly within the range

```
    @Test
    public void testOutOfBounds() {
        try (ColumnView cv = ColumnVector.fromDecimals(
                new BigDecimal("-1E+3"),
                new BigDecimal("1E+3"),
                new BigDecimal("9E+1"),
                new BigDecimal("-9E+1"),
                new BigDecimal("-91"));
             ColumnView expected = ColumnVector.fromBooleans(true, true, false, false, true)) {
            ColumnView result = DecimalUtils.outOfBounds(cv, 1, -1);
            assertColumnsAreEqual(expected, result);
        }
    }
```

**Expected behavior**
The above test should pass

### PR DESCRIPTION
**Pull request by razajafri**
## Description
This PR fixes an error in the `outOfBounds` method in which the `RoundingMode` was selected based on positive values only. The RHS should be rounded towards positive infinity (ROUND_CEILING) for the lower bound and towards negative infinity (ROUND_FLOOR) for the upper bound 

closes #14732

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: java/src/test/java/ai/rapids/cudf/DecimalUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    @Test
+    public void testOutOfBounds() {
+        try (ColumnView cv = ColumnVector.fromDecimals(
+                new BigDecimal("-1E+3"),
+                new BigDecimal("1E+3"),
+                new BigDecimal("9E+1"),
+                new BigDecimal("-9E+1"),
+                new BigDecimal("-91"));
+             ColumnView expected = ColumnVector.fromBooleans(true, true, false, false, true)) {
+            ColumnView result = DecimalUtils.outOfBounds(cv, 1, -1);
```
Comment (ttnghia) [inline java/src/test/java/ai/rapids/cudf/DecimalUtilsTest.java:36]: Memory leak here. This should be put inside the `try()` block.

File: java/src/main/java/ai/rapids/cudf/DecimalUtils.java [hunk 2]
```diff
@@ -82,13 +82,15 @@ public static ColumnVector lessThan(ColumnView lhs, BigDecimal rhs) {
...
+    // Because this is a less-than, and it is rhs that we are rounding, we will round away from 0 (UP) in case rhs is
+    // positive and toward 0 (DOWN) if rhs is negative to make sure we always return the correct value.
     // For example:
-    //      100.1 < 100.19
-    // If we rounded down the rhs 100.19 would become 100.1, and now 100.1 is not < 100.1
-    BigDecimal roundedRhs = rhs.setScale(-leftScale, BigDecimal.ROUND_UP);
+    //      ex:1 100.1 < 100.19
+    //      ex:2 -100.2 < -100.19
+    // In ex:1 If we rounded down the rhs 100.19 would become 100.1, and now 100.1 is not < 100.1
+    // In ex:2 If we rounded up the rhs -100.19 would become -100.2, and now -100.2 is not < -100.2
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/DecimalUtils.java:92]: So what are the correct rounding for these examples? I'm confused. Please be explicit.

File: java/src/main/java/ai/rapids/cudf/DecimalUtils.java [hunk 3]
```diff
@@ -142,7 +144,16 @@ public static ColumnVector greaterThan(ColumnView lhs, BigDecimal rhs) {
...
+
+    // First we have to round the scalar (rhs) to the same scale as lhs.
+    // Because this is a greater-than, and it is rhs that we are rounding, we will round towards 0 (DOWN) in case rhs is
+    // positive and away from 0 (UP) if rhs is negative to make sure we always return the correct value.
+    // For example:
+    //      ex:1 100.2 > 100.19
+    //      ex:2 -100.1 > -100.19
+    // In ex:1 If we rounded up the rhs 100.19 would become 100.2, and now 100.2 is not > 100.2
+    // In ex:2 If we rounded down the rhs -100.19 would become -100.1, and now -100.1 is not > -100.1
+    BigDecimal roundedRhs = rhs.setScale(-cvScale, rhs.signum() > 0 ? BigDecimal.ROUND_DOWN : BigDecimal.ROUND_UP);
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/DecimalUtils.java:156]: This is just repeated of the code above. We should not copy-paste code around. If there is the same code used multiple times, let's extract it into a function and call it multiple times instead.

File: java/src/main/java/ai/rapids/cudf/DecimalUtils.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/DecimalUtils.java:1]: 2024 copyrights

File: java/src/main/java/ai/rapids/cudf/DecimalUtils.java [hunk 4]
```diff
@@ -82,13 +82,15 @@ public static ColumnVector lessThan(ColumnView lhs, BigDecimal rhs) {
     int leftScale = lhs.getType().getScale();
     int leftPrecision = lhs.getType().getDecimalMaxPrecision();
 
-    // First we have to round the scalar (rhs) to the same scale as lhs.  Because this is a
-    // less than and it is rhs that we are rounding, we will round away from 0 (UP)
-    // to make sure we always return the correct value.
+    // First we have to round the scalar (rhs) to the same scale as lhs.
+    // Because this is a less-than, and it is rhs that we are rounding, we will round away from 0 (UP) in case rhs is
+    // positive and toward 0 (DOWN) if rhs is negative to make sure we always return the correct value.
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/DecimalUtils.java:87]: A simpler way to describe this is that we want to round towards positive infinity, i.e.: prefer to move to the next most positive value when throwing away precision.

File: java/src/main/java/ai/rapids/cudf/DecimalUtils.java [hunk 5]
```diff
@@ -142,7 +144,16 @@ public static ColumnVector greaterThan(ColumnView lhs, BigDecimal rhs) {
     // For example:
     //      100.2 > 100.19
     // If we rounded up the rhs 100.19 would become 100.2, and now 100.2 is not > 100.2
-    BigDecimal roundedRhs = rhs.setScale(-cvScale, BigDecimal.ROUND_DOWN);
+
+    // First we have to round the scalar (rhs) to the same scale as lhs.
+    // Because this is a greater-than, and it is rhs that we are rounding, we will round towards 0 (DOWN) in case rhs is
+    // positive and away from 0 (UP) if rhs is negative to make sure we always return the correct value.
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/DecimalUtils.java:150]: Can simplify to state we're rounding towards negative infinity, i.e.: always move to the next lower value when losing precision.

File: java/src/main/java/ai/rapids/cudf/DecimalUtils.java (hunk 3 above)
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/DecimalUtils.java:156]:
```suggestion
    BigDecimal roundedRhs = rhs.setScale(-cvScale, BigDecimal.ROUND_FLOOR);
```

**Comment by razajafri** [comment]
@jlowe @ttnghia Thanks for taking a look PTAL again

File: java/src/main/java/ai/rapids/cudf/DecimalUtils.java (hunk 3 above)
Comment (razajafri) [reply java/src/main/java/ai/rapids/cudf/DecimalUtils.java]: This makes it much better! Thanks

File: java/src/main/java/ai/rapids/cudf/DecimalUtils.java [hunk 6]
```diff
@@ -1,6 +1,6 @@
 /*
  *
- *  Copyright (c) 2022, NVIDIA CORPORATION.
+ *  Copyright (c) 2024, NVIDIA CORPORATION.
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/DecimalUtils.java:3]:
```suggestion
 *  Copyright (c) 2022-2024, NVIDIA CORPORATION.
```

Comment (razajafri) [reply java/src/main/java/ai/rapids/cudf/DecimalUtils.java]: noob!

**Review by jlowe — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

**Pull request merged**